docs/proposals/20200602-machine-deletion-phase-hooks.md

```diff
@@ -0,0 +1,289 @@
+
+- Utilize annotations to implement lifecycle hooks.
+- Each lifecycle point can have 0 or more hooks.
+- Hooks do not enforce ordering.
+- Hooks found during machine reconciliation effectively pause reconciliation
+until all hooks for that lifecycle point are removed from a machine's annotations.
+
+[PlantUML](http://plantuml.com) is the preferred tool to generate diagrams,
+place your `.plantuml` files under `images/` and run `make diagrams` from the docs folder.
+
+### User Stories
+#### Story 1
+(pre-delete) As an operator, I would like to have the ability to perform
+different actions between the time a machine is marked deleted in the api and
+the time the machine is deleted from the cloud.
+
+One use-case is when replacing a control plane machine, ensure a new control
+plane machine has been successfully created and joined to the cluster before
+removing the instance for the deleted machine. This might be useful in case
+there are disruptions during replacement and we need the disk of the existing
+instance to perform some disaster recovery operation.
+
+#### Story 2
+(pre-drain) As an operator, I want the ability to utilize my own draining
+controller instead of the logic built into the machine-controller.  This will
+allow me better flexibility and control over the lifecycle of workloads on each
+node.
+
+#### Story 3
+(pre-drain) As an operator, when I am deleting a control plane machine for
+maintenance reasons, I want to ensure my existing control plane machine is not
+drained until after my replacement comes online.  This will prevent protracted
+periods of a missing control plane member if the replacement machine cannot be
+created in a timely manner.
+
+
+### Implementation Details/Notes/Constraints
+
+For each defined lifecycle point, one or more hooks may be applied as an annotation to the machine object.  These annotations will pause reconciliation of a machine object until all hooks are resolved for that lifecycle point.  The hooks should be managed by an Owning Controller or other external application, or manually created and removed by an administrator.
+
+#### Lifecycle Points
+##### pre-drain
+Hooks defined at this point will prevent the machine-controller from draining a node after the machine-object has been marked for deletion until the hooks are removed.
+##### pre-term
+Hooks defined at this point will prevent the machine-controller from
+removing/terminating the instance in the cloud provider until the hooks are
+removed.
+
+
+#### Annotation Form
+```
+<lifecycle-point>.hook.machine.cluster-api.x-k8s.io/<hook-name>: <owner/creator>
+```
+
+##### lifecycle-point
+This is the point in the lifecycle of reconciling a machine the annotation will have effect and pause the machine-controller.
+
+##### hook-name
+Each hook should have a unique and descriptive name that describes in 1-3 words what the intent/reason for the hook is.  Each hook name should be unique and managed by a single entity.
+
+##### owner (Optional)
```

Comment:
```suggestion
##### owner
```
As per comment above, IMO this should not be optional

Reply:
A well behaved hook should indeed include this information, but it's optional. Existing convention for annotations elsewhere is to respect an annotation if it's merely present.

For instance, consider "some.condition: true".  some.condition is routinely assumed to be 'true' by virtue of being present in the map of annotations, rather than actually inspecting the value component of the annotation.  In many (most?) cases "some.condition: false" would exhibit the same exact behavior.

Reply:
> A well behaved hook should indeed include this information,

That's what's I want to make more explicit in this doc, by removing optional or by adding a note that if the owner is not specified, it increases the risk of name clash among different provider/controllers introducing new annotations
WRT to annotation usage, I'm fine with any convention given that we describe it

Reply:
Yeah, a webhook could be an okay follow-on.  I don't think it's necessary, and I think we should conform to how annotations are generally used where the value portion is optional.  One potential problem is if people start setting the values to 'true' and then someone else might think it's appropriate to edit something to 'false' and expect that something different is supposed to happen.  For me, I like this being clear:  If the annotation is present, regardless of value, we wait.

Reply:
Given the key is what's unique, should we instead say the value is ignored?

Reply:
We can add some wording around that.  It will be ignored by the machine-controller, but it's also meant to be non-nil and provide context to the administrator, so I think saying it's 'ignored' is a little confusing.

Reply:
I think what I'm suggesting is that we consider removing the concept of an owner as the value of the annotation, mainly because it doesn't mean or do anything. If we were able to do something like

- `pre-drain.hook.machine.cluster-api.x-k8s.io: some-hook-owner`
- `pre-drain.hook.machine.cluster-api.x-k8s.io: another-hook-owner`

then I think the value has meaning. But this obviously isn't allowed because you can't have 2 different values for the same key. I'd rather treat the hook name as equivalent to owner name, and use the empty string for the value.

Reply:
I believe the idea behind having the optional owner was to give a way for something other than the controller implementing the hook to set it in a way that a user could go back and track down why it may have been added.

I'm not overly partial either way on it's presence or absence, and if we just take action (or lack of action in the case of pausing reconciliation in the core controllers), then the absence or presence of it doesn't affect the logic.

Reply:
> I believe the idea behind having the optional owner was to give a way for something other than the controller implementing the hook to set it in a way that a user could go back and track down why it may have been added.

Yes, this is the idea.  The hook name (portion after the '/') is supposed to describe what the hook is for, the optional 'owner' portion is mean to describe to the human what controller is supposed to be doing the work.  This way, if someone needs to debug the hook process, they can get a quick hint as to where to go looking.

It's optional because, strictly speaking, it won't be enforced, but it is highly encouraged.

Reply:
👍, let's proceed 😄